File: src/host/video_sender.py
#!/usr/bin/env python3
"""Zynq 以太网视频上位机 — UDP RGB565 推流.

协议: 每包 [u32 LE byte_offset][RGB565 载荷], 载荷 ≤1396B, 板端按 offset 写帧缓.

示例:
  python video_sender.py                          # 内置动画
  python video_sender.py --video demo.mp4         # 真实视频
  python video_sender.py --image logo.png --fps 25
  python video_sender.py --webcam 0
"""
from __future__ import annotations

import argparse
import math
import os
import socket
import struct
import subprocess
import sys
import time
from pathlib import Path

import numpy as np

W, H = 512, 300
FRAME_BYTES = W * H * 2
HDR = 4
MTU_PAYLOAD = 1400 - HDR  # 1396

# 常见 FFmpeg 安装位置（Windows）
_FFMPEG_CANDIDATES = [
    r"C:\Users\wenqu\scoop\apps\ffmpeg\current\bin\ffmpeg.exe",
    r"C:\Users\wenqu\scoop\apps\ffmpeg\9.0.1\bin\ffmpeg.exe",
    r"D:\Software\Ghost\FFmpeg\ffmpeg.exe",
    r"C:\ffmpeg\bin\ffmpeg.exe",
]


def rgb888_to_rgb565(img: np.ndarray) -> bytes:
    """RGB888 uint8 HxWx3 → little-endian RGB565 bytes."""
    r = (img[:, :, 0] >> 3).astype(np.uint16)
    g = (img[:, :, 1] >> 2).astype(np.uint16)
    b = (img[:, :, 2] >> 3).astype(np.uint16)
    pix = (r << 11) | (g << 5) | b
    return pix.astype("<u2").tobytes()


def load_image(path: str | None) -> np.ndarray:
    if path:
        try:
            from PIL import Image

            im = Image.open(path).convert("RGB").resize((W, H), Image.BILINEAR)
            return np.asarray(im, dtype=np.uint8)
        except Exception as e:
            print(f"[WARN] PIL load failed: {e}, using synthetic")
    y = np.linspace(0, 255, H, dtype=np.uint8)[:, None]
    x = np.linspace(0, 255, W, dtype=np.uint8)[None, :]
    img = np.stack([x.repeat(H, 0), y.repeat(W, 1), ((x + y) // 2)], axis=-1)
    return img.astype(np.uint8)


def anim_frame(n: int) -> np.ndarray:
    """内置测试场景：渐变 + 弹球 + 滚动条."""
    t = n * 0.05
    yy, xx = np.mgrid[0:H, 0:W].astype(np.float32)

    g = (np.sin(xx * 0.02 + t) * 0.5 + 0.5) * 180 + 40
    b = (np.cos(yy * 0.025 - t * 0.8) * 0.5 + 0.5) * 180 + 40
    r = (np.sin((xx + yy) * 0.015 + t * 0.6) * 0.5 + 0.5) * 160 + 50
    img = np.stack([r, g, b], axis=-1)

    balls = [
        (80, 180, 28, 255, 60, 60),
        (60, 220, 22, 60, 255, 60),
        (100, 140, 18, 60, 60, 255),
        (45, 280, 14, 255, 255, 60),
    ]
    for i, (rx, ry, rad, cr, cg, cb) in enumerate(balls):
        cx = (W * 0.5) + (W * 0.35) * math.sin(t * (1.1 + i * 0.17) + i)
        cy = (H * 0.5) + (H * 0.32) * math.cos(t * (0.9 + i * 0.13) + i * 1.7)
        dx = xx - cx
        dy = yy - cy
        mask = (dx * dx + dy * dy) < (rad * rad)
        img[mask] = (cr, cg, cb)

    bar = ((xx + n * 6) // 16).astype(np.int32) % 3
    strip = np.zeros((H, W, 3), dtype=np.float32)
    strip[:, :, 0] = np.where(bar == 0, 255, 20)
    strip[:, :, 1] = np.where(bar == 1, 255, 20)
    strip[:, :, 2] = np.where(bar == 2, 255, 20)
    img[H - 24 :, :] = strip[H - 24 :, :]

    ph = (n // 3) % 8
    img[8:28, 8:28] = (30, 30, 30)
    img[12:24, 12 + ph * 2 : 16 + ph * 2] = (255, 255, 0)
    return np.clip(img, 0, 255).astype(np.uint8)


def find_ffmpeg() -> str | None:
    for c in _FFMPEG_CANDIDATES:
        if os.path.isfile(c):
            return c
    for p in os.environ.get("PATH", "").split(os.pathsep):
        cand = os.path.join(p, "ffmpeg.exe")
        if os.path.isfile(cand):
            return cand
        cand = os.path.join(p, "ffmpeg")
        if os.path.isfile(cand):
            return cand
    return None


def _decode_jpeg_bytes(jpg: bytes):
    from io import BytesIO
    from PIL import Image

    im = Image.open(BytesIO(jpg)).convert("RGB")
    if im.size != (W, H):
        im = im.resize((W, H), Image.BILINEAR)
    return np.asarray(im, dtype=np.uint8)


def mjpeg_file_iter(path: str, loop: bool = True):
    from PIL import Image  # noqa: F401

    data = Path(path).read_bytes()
    print(f"[TX] mjpeg file: {path} ({len(data)} bytes)")
    while True:
        i = 0
        n = 0
        while True:
            s = data.find(b"\xff\xd8", i)
            if s < 0:
                break
            e = data.find(b"\xff\xd9", s + 2)
            if e < 0:
                break
            try:
                yield _decode_jpeg_bytes(data[s : e + 2])
                n += 1
            except Exception:
                pass
            i = e + 2
        if not loop:
            break
        print(f"[TX] mjpeg EOF ({n} frames), loop restart")


def ffmpeg_frame_iter(path: str, loop: bool = True):
    ff = find_ffmpeg()
    if not ff:
        raise SystemExit(
            "ffmpeg not found. Install via: winget install Gyan.FFmpeg\n"
            "Or convert to .mjpeg and pass --video file.mjpeg"
        )
    if not os.path.isfile(path):
        raise SystemExit(f"video not found: {path}")

    probe = subprocess.run(
        [ff, "-hide_banner", "-decoders"], capture_output=True, text=True, timeout=20
    )
    if "h264" not in (probe.stdout or ""):
        raise SystemExit(
            "This FFmpeg has no H.264 decoder.\n"
            "Use: winget install Gyan.FFmpeg\n"
            "Or convert to .mjpeg first."
        )

    while True:
        cmd = [
            ff, "-hide_banner", "-loglevel", "error", "-nostdin",
            "-i", path,
            "-an",
            "-vf", f"scale={W}:{H}:flags=lanczos",
            "-f", "mjpeg", "-q:v", "2",
            "-",
        ]
        print(f"[TX] ffmpeg mjpeg pipe: {path}")
        proc = subprocess.Popen(
            cmd, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL, bufsize=1024 * 256
        )
        assert proc.stdout is not None
        buf = b""
        while True:
            chunk = proc.stdout.read(65536)
            if not chunk:
                break
            buf += chunk
            while True:
                s = buf.find(b"\xff\xd8")
                if s < 0:
                    buf = b""
                    break
                e = buf.find(b"\xff\xd9", s + 2)
                if e < 0:
                    buf = buf[s:]
                    break
                jpg = buf[s : e + 2]
                buf = buf[e + 2 :]
                try:
                    yield _decode_jpeg_bytes(jpg)
                except Exception:
                    pass
        proc.stdout.close()
        proc.wait()
        if not loop:
            break
        print("[TX] video EOF, loop restart")


def frame_iter(args):
    if args.video:
        low = args.video.lower()
        if low.endswith((".mjpeg", ".mjpg")):
            yield from mjpeg_file_iter(args.video, loop=not args.once)
        else:
            yield from ffmpeg_frame_iter(args.video, loop=not args.once)
    elif args.webcam is not None:
        try:
            import cv2
        except ImportError:
            raise SystemExit("pip install opencv-python for --webcam")
        cap = cv2.VideoCapture(args.webcam)
        if not cap.isOpened():
            raise SystemExit(f"cannot open webcam index {args.webcam}")
        while True:
            ok, frame = cap.read()
            if not ok:
                break
            frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
            frame = cv2.resize(frame, (W, H), interpolation=cv2.INTER_AREA)
            yield frame
        cap.release()
    elif args.anim:
        n = 0
        while True:
            yield anim_frame(n)
            n += 1
            if args.max_frames and n >= args.max_frames:
                break
    else:
        base = load_image(args.image)
        n = 0
        while True:
            frame = base.copy()
            xs = (n * 4) % max(W - 8, 1)
            frame[:, xs : xs + 8, :] = 255 - frame[:, xs : xs + 8, :]
            yield frame
            n += 1
            if args.max_frames and n >= args.max_frames:
                break


def send_frame(sock: socket.socket, dest: tuple[str, int], payload: bytes) -> int:
    """按 offset 协议发送一帧，返回包数."""
    pkts = 0
    for off in range(0, len(payload), MTU_PAYLOAD):
        chunk = payload[off : off + MTU_PAYLOAD]
        sock.sendto(struct.pack("<I", off) + chunk, dest)
        pkts += 1
    return pkts


def main():
    ap = argparse.ArgumentParser(
        description="Zynq ETH video UDP sender (512x300 RGB565)"
    )
    ap.add_argument("--ip", default="192.168.1.10", help="board PL IP")
    ap.add_argument("--port", type=int, default=5001)
    ap.add_argument(
        "--src",
        default="192.168.1.100",
        help="PC NIC bind IP; empty string to use default route",
    )
    ap.add_argument("--image", default=None)
    ap.add_argument("--anim", action="store_true")
    ap.add_argument("--video", default=None)
    ap.add_argument("--webcam", type=int, default=None)
    ap.add_argument("--once", action="store_true")
    ap.add_argument("--fps", type=float, default=30.0)
    ap.add_argument("--count", type=int, default=0, help="stop after N frames")
    ap.add_argument(
        "--max-frames",
        type=int,
        default=0,
        help="generator limit for --anim/image (0=unlimited)",
    )
    ap.add_argument("--quiet", action="store_true", help="less log")
    args = ap.parse_args()

    if args.image is None and args.video is None and args.webcam is None:
        args.anim = True
    if args.count:
        args.max_frames = args.max_frames or args.count

    dest = (args.ip, args.port)
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_SNDBUF, 1 << 20)
    if args.src:
        try:
            sock.bind((args.src, 0))
            print(f"[TX] bind {args.src}")
        except OSError as e:
            print(f"[WARN] bind {args.src} failed: {e}, using default route")

    period = 1.0 / max(args.fps, 0.1)
    if args.video:
        mode = "video"
    elif args.webcam is not None:
        mode = "webcam"
    elif args.anim:
        mode = "anim"
    else:
        mode = "image"
    print(f"[TX] {args.ip}:{args.port} {W}x{H} RGB565 @ {args.fps:.1f} fps mode={mode}")

    n = 0
    t0 = time.perf_counter()
    next_t = t0
    try:
        for frame in frame_iter(args):
            if frame.shape[0] != H or frame.shape[1] != W:
                frame = np.ascontiguousarray(frame[:H, :W])
            payload = rgb888_to_rgb565(frame)
            if len(payload) != FRAME_BYTES:
                print(f"[ERR] bad frame size {len(payload)}")
                continue
            send_frame(sock, dest, payload)
            n += 1
            if n == 1 and not args.quiet:
                pkts = (len(payload) + MTU_PAYLOAD - 1) // MTU_PAYLOAD
                print(f"[TX] first frame {len(payload)} bytes ({pkts} pkts)")
            if n % 30 == 0 and not args.quiet:
                dt = time.perf_counter() - t0
                print(f"[TX] frames={n} ~{n / max(dt, 0.001):.1f} fps")
            if args.count and n >= args.count:
                break
            # 定时发送：补偿处理耗时，减少抖动
            next_t += period
            delay = next_t - time.perf_counter()
            if delay > 0:
                time.sleep(delay)
            else:
                next_t = time.perf_counter()
    except KeyboardInterrupt:
        print("\n[TX] stop")
    finally:
        sock.close()
        dt = time.perf_counter() - t0
        print(f"[TX] sent {n} frames in {dt:.2f}s (~{n / max(dt, 0.001):.1f} fps)")


if __name__ == "__main__":
    main()

File: src/host/video_sender.mjs
#!/usr/bin/env node
/**
 * video_sender.mjs — 与 host/video_sender.py 同协议的 UDP 推流（无需 numpy/python）
 *
 * 协议: 每包 [u32 LE byte_offset][RGB565 载荷], 载荷 <= 1392 B（8 的倍数）
 * 用法: node video_sender.mjs [--ip 192.168.1.10] [--port 5001]
 *                             [--src 192.168.1.100] [--fps 15]
 *                             [--pace-mpbps 15] [--no-pace] [--count N]
 *                             [--test bars|grad|edge]
 *
 * --test 图案：
 *   bars   8px 横向条纹整体滚动（黑横纹一眼可见）+ 每帧右移的黄色方块（拖影判据）
 *          + 左边缘 8 列奇偶行洋红/黑标记（行序错乱判据）
 *   grad   缓变渐变（查量化/色带）
 *   edge   整屏逐帧黑白交替（查换帧是否原子：非原子会看到灰行或残影）
 */
import dgram from 'node:dgram';
import { writeFileSync } from 'node:fs';

const W = 512, H = 300, FRAME_BYTES = W * H * 2, HDR = 4;
// 载荷必须是 8 的倍数：见 video_sender.py 里 MTU_PAYLOAD 的注释（否则每包边界会毁掉一个 64bit DDR 字）
// --mtu-payload 1396 是用来**复现**这个错误的（A/B 对照实验用），不是让你日常这么发。
const MTU = Number(get('mtu-payload', 1392));
if (MTU % 8) console.log(`[TX] 警告：MTU_PAYLOAD=${MTU} 不是 8 的倍数，包边界会毁掉 64bit 字（规律黑点）`);

function get(name, def) {
  const i = process.argv.indexOf('--' + name);
  if (i < 0) return def;
  const nxt = process.argv[i + 1];
  return nxt && !nxt.startsWith('--') ? nxt : true;
}
const IP = String(get('ip', '192.168.1.10'));
const PORT = Number(get('port', 5001));
const SRC = String(get('src', '192.168.1.100'));
const FPS = Number(get('fps', 15));
const NO_PACE = get('no-pace', false) === true;
const PACE = NO_PACE ? 0 : Number(get('pace-mpbps', 15));
const COUNT = Number(get('count', 0));
const TEST = String(get('test', 'bars'));
// --dump <文件>：把最后两帧写到 <文件> 和 <文件>.prev，供 host/ddr_verify.mjs --ref 逐字节比对
const DUMP = typeof get('dump', null) === 'string' ? String(get('dump', '')) : null;
let lastFrame = null;

const buf = Buffer.alloc(FRAME_BYTES);

function put(x, y, r, g, b) {
  buf.writeUInt16LE(((r >> 3) << 11) | ((g >> 2) << 5) | (b >> 3), (y * W + x) * 2);
}

function build(n) {
  if (TEST === 'move') {
    // 一次同时判两件事：
    //  · 方块只在白象限内部往返（永不触边、永不跨参考线）⇒ 看到重影/拖尾就是换帧不原子
    //  · 蓝象限铺 4 像素粗网格 ⇒ 若粗网格在右屏不再闪烁，说明之前的细线闪烁是
    //    最近邻缩放的走样（1px 线落到采样间隙），不是数据问题
    const bx = 270 + (1 - Math.cos(n * 0.10)) * 70;   // 270..410，白象限内往返
    const b0 = Math.round(bx);
    for (let y = 0; y < H; y++) {
      for (let x = 0; x < W; x++) {
        let rgb;
        if (y < 150) rgb = (x < 256) ? [255, 0, 0] : [0, 255, 0];
        else         rgb = (x < 256) ? [0, 0, 255] : [255, 255, 255];
        if (y >= 160 && y < 300 && x >= 8 && x < 248 &&
            ((x % 32) < 4 || (y % 32) < 4)) rgb = [255, 255, 255];
        if (x < 2 || x >= W - 2 || y < 2 || y >= H - 2) rgb = [255, 255, 255];
        if (x === 256 || x === 257 || y === 150 || y === 151) rgb = [0, 0, 0];
        if (x >= b0 && x < b0 + 50 && y >= 190 && y < 240) rgb = [255, 0, 0];
        put(x, y, rgb[0], rgb[1], rgb[2]);
      }
    }
    return buf;
  }
  if (TEST === 'frameid') {
    // 每个 16bit 像素 = (字号 + 帧号) ⇒ 回读 DDR 就能反解「这个字是第几帧写进去的」：
    // n_implied = (lane - w) & 0xffff。用来定量测换帧滞后与逐字混合程度。
    for (let w = 0; w < FRAME_BYTES / 8; w++) {
      const v = (w + n) & 0xffff;
      const o = w * 8;
      buf.writeUInt16LE(v, o); buf.writeUInt16LE(v, o + 2);
      buf.writeUInt16LE(v, o + 4); buf.writeUInt16LE(v, o + 6);
    }
    return buf;
  }
  if (TEST === 'hold') {
    // 内容完全与帧号无关（每帧一模一样）+ 方块固定居中：
    //   · 若屏幕上方块仍撕裂/带残影 ⇒ 显示数据通路问题（两帧相同不可能混出差异）
    //   · 若干净 ⇒ 之前的“撕裂”来自图案自身被裁切 / 或只在运动时才有换帧混合
    // 蓝象限里再铺 1 像素白网格（每 32 像素），用来查行级错位。
    for (let y = 0; y < H; y++) {
      for (let x = 0; x < W; x++) {
        let rgb;
        if (y < 150) rgb = (x < 256) ? [255, 0, 0] : [0, 255, 0];
        else         rgb = (x < 256) ? [0, 0, 255] : [255, 255, 255];
        if (y >= 160 && y < 300 && x >= 8 && x < 248 && (x % 32 === 0)) rgb = [255, 255, 255];
        if (y >= 160 && y < 300 && x >= 8 && x < 248 && (y % 32 === 0)) rgb = [255, 255, 255];
        if (x < 2 || x >= W - 2 || y < 2 || y >= H - 2) rgb = [255, 255, 255];
        if (x === 256 || x === 257 || y === 150 || y === 151) rgb = [0, 0, 0];
        if (x >= 128 && x < 188 && y >= 200 && y < 260) rgb = [255, 255, 0];
        put(x, y, rgb[0], rgb[1], rgb[2]);
      }
    }
    return buf;
  }
  if (TEST === 'blocks') {
    // 大色块 + 单像素参考线：纯色场不会产生手机摩尔纹，所以屏幕上看到任何
    // 椒盐点都是真实数据问题；同时用来判断换帧是否原子、有没有拖影。
    for (let y = 0; y < H; y++) {
      for (let x = 0; x < W; x++) {
        let rgb;
        if (y < 150) rgb = (x < 256) ? [255, 0, 0] : [0, 255, 0];
        else         rgb = (x < 256) ? [0, 0, 255] : [255, 255, 255];
        if (x < 2 || x >= W - 2 || y < 2 || y >= H - 2) rgb = [255, 255, 255]; // 边框
        if (x === 256 || x === 257) rgb = [0, 0, 0];                          // 竖参考线
        if (y === 150 || y === 151) rgb = [0, 0, 0];                          // 横参考线
        const cx = ((n * 6) % (W + 120)) - 60;
        if (x >= cx && x < cx + 60 && y >= 200 && y < 260) rgb = [255, 255, 0];
        put(x, y, rgb[0], rgb[1], rgb[2]);
      }
    }
    return buf;
  }
  if (TEST === 'wordid') {
    // 自描述图案：每个 64bit DDR 字（4 像素）填它自己的字号（可加偏移做相位位移，
    // 用来区分「这一帧真丢了」和「上一帧碰巧写过同样的值」）
    const add = Number(get('wordid-add', 0)) & 0xffff;
    for (let w = 0; w < FRAME_BYTES / 8; w++) {
      const v = (w + add) & 0xffff;
      const o = w * 8;
      buf.writeUInt16LE(v, o); buf.writeUInt16LE(v, o + 2);
      buf.writeUInt16LE(v, o + 4); buf.writeUInt16LE(v, o + 6);
    }
    return buf;
  }
  const phase = n % 64;
  const alt = (n & 1) ? 255 : 0;
  const cx = ((n * 7) % (W + 120)) - 60;
  for (let y = 0; y < H; y++) {
    for (let x = 0; x < W; x++) {
      if (TEST === 'edge') {
        put(x, y, alt, alt, alt);
      } else if (TEST === 'grad') {
        put(x, y, (x + n * 2) & 255, (y * 2) & 255, 128);
      } else {
        if (((y + phase) >> 3) & 1) put(x, y, 255, 255, 255);
        else put(x, y, 0, 90, 200);
        if (x < 8) { if ((y >> 1) & 1) put(x, y, 255, 0, 255); else put(x, y, 0, 0, 0); }
        if (x >= cx && x < cx + 60 && y >= 140 && y < 200) put(x, y, 255, 255, 0);
      }
    }
  }
  return buf;
}

const sock = dgram.createSocket('udp4');
if (SRC) {
  try { sock.bind({ address: SRC, port: 0 }); }
  catch (e) { console.log(`[WARN] bind ${SRC} failed: ${e.message}, use default route`); }
}

// 绝对时间限速：帧内逐包匀速发出，避免 221 包以线速倾泻打爆板端入包 FIFO。
// 15 MB/s 下每包只有 ~93 µs 预算，Date.now() 的 1 ms 精度不够，用 performance.now()。
const _sab = new Int32Array(new SharedArrayBuffer(4));
const nowMs = () => performance.now();
function sleepMs(ms) { Atomics.wait(_sab, 0, 0, ms); }

let paceNext = nowMs();
function paceSpend(bytes) {
  if (!PACE) return;
  const target = paceNext + (bytes / (PACE * 1e6)) * 1000;
  if (target < nowMs()) { paceNext = nowMs(); return; }
  const wait = target - nowMs();
  if (wait > 2) sleepMs(wait - 1.5);
  while (nowMs() < target) { /* tail spin */ }
  paceNext = target;
}

function sendFrame(payload) {
  let pkts = 0;
  for (let off = 0; off < payload.length; off += MTU) {
    const chunk = payload.subarray(off, Math.min(off + MTU, payload.length));
    const p = Buffer.allocUnsafe(chunk.length + HDR);
    p.writeUInt32LE(off, 0);
    chunk.copy(p, HDR);
    sock.send(p, 0, p.length, PORT, IP);
    paceSpend(p.length);
    pkts++;
  }
  return pkts;
}

console.log(`[TX] ${IP}:${PORT} ${W}x${H} RGB565 @${FPS}fps test=${TEST} pace=${PACE} MB/s src=${SRC || 'default'}`);
const period = 1000 / FPS;
let n = 0, t0 = Date.now(), nextT = t0;

function tick() {
  build(n);
  sendFrame(buf);
  n++;
  if (n === 1) console.log(`[TX] first frame ${FRAME_BYTES} B = ${Math.ceil(FRAME_BYTES / MTU)} pkts`);
  if (DUMP) {                       // 保留最后两帧：乒乓两个 bank 各对应其一
    try { writeFileSync(DUMP + '.prev', lastFrame); } catch (e) {}
    lastFrame = Buffer.from(buf);
  }
  if (n % 30 === 0) {
    const dt = (Date.now() - t0) / 1000;
    console.log(`[TX] frames=${n} ~${(n / dt).toFixed(1)} fps`);
  }
  if (COUNT && n >= COUNT) {
    // 等内核把已排队的 sendto 真正发出去再关 socket，
    // 否则 --count 1~2 这种短测会把大部分包丢在发送队列里。
    console.log('[TX] waiting for the tx queue to drain ...');
    setTimeout(() => {
      if (DUMP && lastFrame) { try { writeFileSync(DUMP, lastFrame); } catch (e) { console.log('[TX] dump failed: ' + e.message); } }
      sock.close();
    }, 2000);
    return;
  }
  nextT += period;
  const d = nextT - Date.now();
  setTimeout(tick, d > 0 ? d : 0);
  if (d <= 0) nextT = Date.now();
}
tick();

process.on('SIGINT', () => { console.log(`\n[TX] sent ${n} frames`); sock.close(); process.exit(0); });
